tempesta_fw/t/unit/test_frang.c

```diff
@@ -0,0 +1,394 @@
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * his program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include <linux/inet.h>
+
+#include "../../gfsm.h"
+#include "../../http.h"
+#include "../../log.h"
+
+#include "helpers.h"
+#include "kallsyms_helper.h"
+#include "test.h"
+
+#define FRANG_HASH_BITS 17
+#define FRANG_FREQ	8
+#define HANDLER_OFF	0
+
+typedef struct {
+	unsigned long	ts;
+	unsigned int	conn_new;
+	unsigned int	req;
+} FrangRates;
+
+typedef struct frang_account_t {
+	struct hlist_node hentry;
+	struct in6_addr addr;	/* client address */
+	unsigned long last_ts;	/* last access time */
+	unsigned int conn_curr;	/* current connections number */
+	FrangRates history[FRANG_FREQ];
+} FrangAcc;
+
+typedef struct {
+	char *str;
+	size_t len;		/* The pre-computed strlen(@str). */
+} FrangCtVal;
+
+typedef struct {
+	/* Limits (zero means unlimited). */
+	unsigned int req_rate;
+	unsigned int req_burst;
+	unsigned int conn_rate;
+	unsigned int conn_burst;
+	unsigned int conn_max;
+	/*
+	Limits on time it takes to receive
+	 a full header or a body chunk.
+	*/
+	unsigned long clnt_hdr_timeout;
+	unsigned long clnt_body_timeout;
+	/* Limits for HTTP request contents: uri, headers, body, etc. */
+	unsigned int http_uri_len;
+	unsigned int http_field_len;
+	unsigned int http_body_len;
+	unsigned int http_hchunk_cnt;
+	unsigned int http_bchunk_cnt;
+	bool http_ct_required;
+	bool http_host_required;
+	/* The bitmask of allowed HTTP Method values. */
+	unsigned long http_methods_mask;
+	/* The list of allowed Content-Type values. */
+	FrangCtVal *http_ct_vals;
+} FrangCfg;
+
+const int (*frang_conn_new) (struct sock *);
+int (*frang_http_req_handler)(void *obj,struct sk_buff *skb, 
+			      unsigned int off);
+
+struct inet_sock *isk;
+struct sock mocksock;
+const char *inet_addr = "192.168.245.128";
+FrangCfg *frang_cfg;
+
+static TfwConnection *
+test_conn_alloc(void)
+{
+	TfwConnection *conn;
+	static struct kmem_cache *test_conn_cache = NULL;
+	if(test_conn_cache == NULL)
+		test_conn_cache = kmem_cache_create("tfw_test_conn_cache",
+				        sizeof(TfwConnection), 0, 0, NULL);
+	BUG_ON(test_conn_cache == NULL);
+	conn = kmem_cache_alloc(test_conn_cache, GFP_ATOMIC);
+	BUG_ON(!conn);
+	tfw_connection_init(conn);
+	return conn;
+}
+
+static int
+req_handler(TfwHttpReq  *req)
+{
+	TfwConnection *conn;
+
+	conn = test_conn_alloc();
+	conn->msg = &req->msg;
+	conn->sk = &mocksock;
+	isk = (struct inet_sock *) (&mocksock);
+	isk->inet_saddr = htonl(in_aton(inet_addr));
+	if (!conn->sk->sk_security) {
+		frang_conn_new(conn->sk);
+	}
+
+	return frang_http_req_handler((void *) conn,
+					       req->msg.skb_list.first, HANDLER_OFF);
+}
+
+static TfwHttpReq *
+get_test_req(const char *req)
+{
+	TfwHttpReq *test_req;
+	static char req_str_copy[PAGE_SIZE]; 
+	int len = strlen(req);
+	BUG_ON(len+1 > sizeof(req_str_copy));
```

Comment:
I write "Add BUG_ON(len + 1 > sizeof(req_str_copy));"
Add is add, not 'change', 'edit' or something else